src/app/AttributeAccessInterfaceCache.h

```diff
@@ -0,0 +1,145 @@
+        {
+            entry.Invalidate();
+        }
+        mLastUnusedEntry.Invalidate();
+    }
+
+    /**
+     * @brief Mark that we know a given <`endpointId`, `clusterId`> uses AAI, with instance `attrInterface`
+     */
+    void MarkUsed(EndpointId endpointId, ClusterId clusterId, AttributeAccessInterface * attrInterface)
+    {
+        GetCacheSlot(endpointId, clusterId)->Set(endpointId, clusterId, attrInterface);
+    }
+
+    /**
+     * @brief Mark that we know a given <`endpointId`, `clusterId`> does NOT use AAI.
+     */
+    void MarkUnused(EndpointId endpointId, ClusterId clusterId) { mLastUnusedEntry.Set(endpointId, clusterId, nullptr); }
+
+    /**
+     * @brief Get the AttributeAccessInterface instance for a given <`endpointId`, `clusterId`>, if present in cache.
+     *
+     * @param endpointId - Endpoint ID to look-up.
+     * @param clusterId - Cluster ID to look-up.
+     * @param outAttributeAccess - If not null, and Get returns `kDefinitelyUsed`, then this is set to the instance pointer.
+     * @return a for whether the entry is actually used or not.
+     */
+    CacheResult Get(EndpointId endpointId, ClusterId clusterId, AttributeAccessInterface ** outAttributeAccess)
+    {
+        if (mLastUnusedEntry.Matches(endpointId, clusterId))
+        {
+            return CacheResult::kDefinitelyUnused;
+        }
+
+        AttributeAccessCacheEntry * cacheSlot = GetCacheSlot(endpointId, clusterId);
+        if (cacheSlot->Matches(endpointId, clusterId) && (cacheSlot->accessor != nullptr))
+        {
+            if (outAttributeAccess != nullptr)
+            {
+                *outAttributeAccess = cacheSlot->accessor;
+            }
+            return CacheResult::kDefinitelyUsed;
+        }
+
+        return CacheResult::kCacheMiss;
+    }
+
+private:
+    struct AttributeAccessCacheEntry
+    {
+        EndpointId endpointId               = kInvalidEndpointId;
+        ClusterId clusterId                 = kInvalidClusterId;
+        AttributeAccessInterface * accessor = nullptr;
+
+        void Invalidate()
+        {
+            endpointId = kInvalidEndpointId;
+            clusterId  = kInvalidClusterId;
+            accessor   = nullptr;
+        }
+
+        void Set(EndpointId theEndpointId, ClusterId theClusterId, AttributeAccessInterface * theAccessor)
+        {
+            endpointId = theEndpointId;
+            clusterId  = theClusterId;
+            accessor   = theAccessor;
+        }
+
+        bool Matches(EndpointId theEndpointId, ClusterId theClusterId) const
+        {
+            return (endpointId == theEndpointId) && (clusterId == theClusterId);
+        }
+    };
+
+    AttributeAccessCacheEntry * GetCacheSlot(EndpointId endpointId, ClusterId clusterId)
+    {
+        (void) endpointId;
+        (void) clusterId;
+        return &mCacheSlots[0];
+    }
+
+    AttributeAccessCacheEntry mCacheSlots[1];
+    AttributeAccessCacheEntry mLastUnusedEntry;
```

Comment:
mLastUnusedEntry could probably be a ConcreteClusterPath and save one word of RAM... Either way, though.

Reply:
Then I would have to change the logic to invalidate, to be less regular. I will leave this one as-is.